keps/sig-storage/20190530-pv-health-monitor.md

```diff
@@ -0,0 +1,571 @@
+        // +optional
+        Phase PersistentVolumeClaimPhase
+        // AccessModes contains all ways the volume backing the PVC can be mounted
+        // +optional
+        AccessModes []PersistentVolumeAccessMode
+        // Represents the actual resources of the underlying volume
+        // +optional
+        Capacity ResourceList
+        // +optional
+        Conditions []PersistentVolumeClaimCondition
+        // +optional
+        VolumeHealthName *string
+}
+```
+
+Alternative option 2 does not address the concerns from option 1. Assuming the node-level sidecar is running in a pod as a service account, we need to grant that service account permission to modify any instance of the volume health CR, and then the monitoring controller just copies that into the PVC status.
+
+#### Alternative option 3
+
+We can also reuse the PV Taints and introduce a new Taint called PVUnhealthMessage for PV health condition whose key is specific (PVUnhealthMessage) and value can be set differently. 
+
+For example, if the PV is not attached now, we can mark the PV using the PVUnhealthMessage taint like this:
+```
+Key: “PVUnhealthMessage”
+Value: “AttachError - the pv is not attached to node1 now”
+VolumeTaintEffect: NoEffect
+```
+
+If the volume is deleted, we can mark the PV using the PVUnhealthMessage taint like this:
+```
+Key: “PVUnhealthMessage” 
+Value: “VolumeError - the volume is deleted from backend”
+VolumeTaintEffect: NoEffect
+```
+
+Note that:
+
+- all the VolumeTaintEffects are NoEffect now at first, we may talk about the reactions later in another proposal;
+- the taint Value is string now, it is theoretically possible that several errors are detected for one PV, we may extend the string to cover this situation: combine the errors together and splited by semicolon or other symbols.
+
+This was initially brought up to address the external data populator use case, but there were lots of concerns about this proposal. We do not want our main proposal to depend on something that is not approved.
+
+#### Optional HTTP(RPC) service
+
+Create a HTTP(RPC) service to receive volume health condition reports from other components.
+
+Users can extend the volume health condition monitoring by setting up their own detector and report the result to the service.
+
+This optional service is out of scope for this KEP but may be introduced in a future KEP if needed.
+
+This option is not in the main proposal because it is a push-based method while CSI uses pull-based method.
+
+## Graduation Criteria
+### Alpha -> Beta
+* Feature complete, including:
+  * External controller volume health monitoring.
+  * External node agent volume health monitoring.
+* Tests outlined in design proposal implemented.
+
+### Beta -> GA
+* Volume health support is added to multiple CSI drivers.
+* Volume health feature deployed in production and have gone through at least one K8s upgrade.
+
+## Test Plan
+### Unit tests
```

Comment:
It would be nice to be more specific about the test cases, what error scenarios will we be testing? What drivers are we going to use to test?

Also, any plan for stress/scale tests?

Reply:
Updated.